File: src/ar/edu/unlu/sisops/RoundRobin.java
import java.lang.reflect.Array;
import java.util.ArrayList;
import java.util.Scanner;

public class RoundRobin {

    static class Proceso{
        int id;
        int tiempoDeServicio;
        int tiempoRestante;
        int tiempoDeRetorno; //Tiempo total. desde que entró hasta que salió
        int tiempoDeEspera; //Para calcular el tiempo de espera. Tiempo en que no fue atendido

        public Proceso(int id, int tiempoDeServicio){
            this.id = id;
            this.tiempoDeServicio= tiempoDeServicio;
            this.tiempoRestante = tiempoDeServicio;
        }
    }

    static class Registro{
        Proceso proceso;
        float tiempoDeRetorno;
        float tiempoDeEspera;
    }



    private static final float q = 4;
    private static ArrayList<Proceso> procesos = new ArrayList<>();

    public static void agregarProceso(int id, int tiempoDeServicio){
        Proceso p = new Proceso(id, tiempoDeServicio);
        procesos.add(p);
    }
    public static void roundRobin(){
        float tiempoDeIntercambio = q/4;
        float reloj=0;
        ArrayList<Registro> registros = new ArrayList<>();
        Registro registro = new Registro();
        int sumaTiempoRestante = procesos.stream().mapToInt(p ->p.tiempoRestante).sum();
        Proceso procesoAnt = null;
        while(sumaTiempoRestante > 0){
            for (Proceso proceso: procesos) {
                //Si al proceso actual tiene más que Q
                if(proceso.tiempoRestante>q){
                    //Si el proceso es distinto del anterior y si el anterior no es nulo
                    if(procesoAnt != proceso && procesoAnt != null){

                        reloj += tiempoDeIntercambio/2;
                    }
                    //Si el proceso.tiempoRestante es mayor a 4
                    if(proceso.tiempoRestante>q){

                        reloj += tiempoDeIntercambio/2;
                    }

                    //proceso el proceso, sumo al reloj y al tiempo restante
                    proceso.tiempoRestante -= q;
                    reloj += q;

                //else el proceso actual tiene menos que Q
                }else {
                    //Si el proceso es distinto del proceso anterior
                    if(proceso != procesoAnt){
                        reloj += tiempoDeIntercambio/2;
                    }

                    reloj += proceso.tiempoRestante;
                    proceso.tiempoRestante = 0;

                    System.out.println("proceso:"+proceso.id+"ha sido terminado"+"\nReloj:"+reloj);
                    //almaceno en registro el proceso junto con su tiempo de espera y su tiempo de retorno
                    registro.proceso = proceso;
                    registro.tiempoDeEspera = reloj - proceso.tiempoDeServicio;
                    registro.tiempoDeRetorno = reloj;
                    registros.add(registro);
                }
                procesoAnt = proceso;
            }
            sumaTiempoRestante = procesos.stream().mapToInt(p ->p.tiempoRestante).sum();
        }

        //Calcular tiempo de espera promedio
        double tiempoDeEsperaPromedio = registros.stream().mapToDouble(r -> r.tiempoDeEspera).average().getAsDouble();
        double tiempoDeRetornoPromedio = registros.stream().mapToDouble(r -> r.tiempoDeRetorno).average().getAsDouble();

        //Mostrar los tiempos promedios
        System.out.println("Tiempo de espera promedio:"+tiempoDeEsperaPromedio+"\nTiempo de retorno promedio:"+tiempoDeRetornoPromedio);

    }

    public static void main(String[] args){
        Scanner sc = new Scanner(System.in);
        boolean exit = false;


        while(!exit){
            System.out.println("Ingrese una opción:");
            System.out.println("1.Comenzar RoundRobin (con los procesos cargados)");
            System.out.println("2.Cargar más procesos");
            System.out.println("3.Salir");
            int op = sc.nextInt();
            switch (op) {
                case 1 -> {
                    roundRobin();
                    //borrar procesos
                }
                case 2 -> {
                    System.out.println("Ingrese id del proceso:");
                    int id = sc.nextInt();
                    System.out.println("Ingrese tiempo de servicio del proceso:");
                    int tiempoDeServicio = sc.nextInt();
                    agregarProceso(id, tiempoDeServicio);
                }
                case 3 -> {
                    exit = true;
                }
                default -> System.out.println("Ingrese un valor válido.");
            }

        }
    }
}

File: src/ar/edu/unlu/sisops/roundrobin/roundrobin.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct Proceso {
    int PID;
    double tiempoDeServicio;
    double tiempoRestante;
} proceso;

typedef struct Registro {
    int PID;
    double tiempoDeEspera;
    double tiempoDeRetorno;
} registro;

int sumaTiempoRestante(proceso procesos[], int cantProcesos) {
    int suma = 0;
    for (int i = 0; i < cantProcesos; i++) {
        suma += procesos[i].tiempoRestante;
    }
    return suma;
}

void roundRobin(proceso procesos[], registro registros[], int cantProcesos, int quantum) {
    double reloj = 0;
    double tiempoDeIntercambio = quantum / 4;
    int cantRegistros = 0;

    proceso *procesoAnt = NULL;

    do {

        for (int i = 0; i < cantProcesos; i++) {
        
            // si ya se terminó de atender, se sigue con el próximo proceso
            if (procesos[i].tiempoRestante == 0) {
                continue;
            }

            // si no es el primer proceso, o es diferente al anterior, se desaloja
            if (procesoAnt != NULL && &procesos[i] != procesoAnt) {
                reloj += tiempoDeIntercambio / 2;
            }

            // si el tiempo restante es mayor al quantum, se asigna al procesador
            if (procesos[i].tiempoRestante > quantum) {
                reloj += tiempoDeIntercambio / 2;

                // Se atiende el proceso, sumo Q al reloj y resto Q al tiempo restante
                reloj += quantum;
                procesos[i].tiempoRestante -= quantum;
            } 
            // El tiempo restante del proceso actual es menor que Q, se pone en 0 el tiempo restante
            // No se hace el intercambio
            else {
                reloj += procesos[i].tiempoRestante;
                procesos[i].tiempoRestante = 0;

                printf("\nProceso %d terminado", procesos[i].PID);
                printf("\nReloj: %.2f \n", reloj);

                // Almaceno en registros el proceso junto con su tiempo de espera y su tiempo de retorno
                registros[cantRegistros].PID = procesos[i].PID;
                registros[cantRegistros].tiempoDeEspera = reloj - procesos[i].tiempoDeServicio;
                registros[cantRegistros].tiempoDeRetorno = reloj;
                cantRegistros++;
            }
            procesoAnt = &procesos[i];
        }

    // mientras haya procesos por atender
    } while (sumaTiempoRestante(procesos, cantProcesos) > 0);
}

void mostrarResultados(registro registros[], int cantRegistros) {
    double tiempoDeEsperaPromedio = 0, tiempoDeRetornoPromedio = 0;

    for (int i = 0; i < cantRegistros; i++) {
        tiempoDeEsperaPromedio += registros[i].tiempoDeEspera;
        tiempoDeRetornoPromedio += registros[i].tiempoDeRetorno;
    }

    tiempoDeEsperaPromedio /= cantRegistros;
    tiempoDeRetornoPromedio /= cantRegistros;

    printf("--------------------------------- \n");
    printf("           RESULTADOS \n");
    printf("--------------------------------- \n");
    printf("Tiempo de espera promedio: %.2f \n", tiempoDeEsperaPromedio);
    printf("Tiempo de retorno promedio: %.2f \n", tiempoDeRetornoPromedio);
}

void imprimirInicioRoundRobin() {
        printf("\n---------------------------------\n");
        printf("       INICIO ROUND ROBIN \n");
        printf("---------------------------------\n");
    }

void imprimirFinRoundRobin() {
    printf("\n---------------------------------\n");
    printf("        FIN ROUND ROBIN \n");
    printf("---------------------------------\n\n");
}

int main(int argc, char **argv) {

    // argv[0] es el nombre del programa, arg[1] es el quantum
    int quantum = atoi(argv[1]);
    int cantProcesos = argc - 2;

    registro *registros = malloc(sizeof(registro) * cantProcesos);

    proceso *procesos = malloc(sizeof(proceso) * cantProcesos);
    for (int i = 0; i < cantProcesos; i++) {
        procesos[i].PID = i + 1;
        procesos[i].tiempoDeServicio = atoi(argv[i + 2]);
        procesos[i].tiempoRestante = procesos[i].tiempoDeServicio;
    }

    imprimirInicioRoundRobin();

    roundRobin(procesos, registros, cantProcesos, quantum);

    imprimirFinRoundRobin();

    mostrarResultados(registros, cantProcesos);

    return 0;
}

